compiler/include/stlUtil.h

```diff
@@ -34,31 +34,50 @@
 // FnSymbol* fn; and returns one FnSymbol* from the global list of
 // function symbol(pointer)s on each iteration.
 //
+// Implementation Note:
 // To implement iteration compatibly (with, for example, forv_Vec), we
 // need both an index (iterator) and the value obtained by
 // dereferencing it.  Since it is not possible to declare variables
-// with different base types in a for loop, we declare the hidden
-// iterator internally, and push the declaration of the value outside
-// the loop.  That is, the client code has to provide the declaration
-// where formerly it did not.
+// with different base types in a for loop, we declare unique-ified
+// temporary variables to store a reference to the collection, the
+// begin iterator, and the end iterator. The loop still declares
+// the VAL argument since its name cannot be unique-ified. In this
+// way, the implementation takes care to evaluate VEC only once
+// and allows multiple loops using the same index variable.
 //
-// CAVEAT: Using the same name for the value in the macro may lead to
-// compilation errors if both macro invocations are enclosed in the
-// same scope.
+// NOTE: We should consider replacing these with something else.
+// In C++11 perhaps these macros could be written
+//  for(TYPE* VAL: VEC)
 //
-// NOTE: We should consider replacing these with BOOST_FOREACH once
-// more of the Boost libraries are accepted into the C++ standard.
-//
-#define for_vector(TYPE, VAL, VEC) TYPE* VAL;                 \
-  for (std::vector<TYPE*>::iterator _i_##VAL = (VEC).begin(); \
-       (VAL = (_i_##VAL != (VEC).end()) ? *_i_##VAL : (TYPE*)0) ; _i_##VAL++ )
 
-#define for_set(TYPE, VAL, VEC) TYPE* VAL;                 \
-  for (std::set<TYPE*>::iterator _i_##VAL = (VEC).begin(); \
-       (VAL = (_i_##VAL != (VEC).end()) ? *_i_##VAL : (TYPE*)0) ; _i_##VAL++ )
+#define STL_UTIL_TOK_PASTE_INTERNAL(a, b) a ## b
+#define STL_UTIL_TOK_PASTE(a, b) STL_UTIL_TOK_PASTE_INTERNAL(a,b)
+#define STL_UTIL_UNIQUE(NAME) STL_UTIL_TOK_PASTE(NAME, __LINE__)
+
+#define for_vector(TYPE, VAL, VEC) \
+  const std::vector<TYPE*> & STL_UTIL_UNIQUE(for_vector_vec) = (VEC); \
+  std::vector<TYPE*>::const_iterator STL_UTIL_UNIQUE(for_vector_iter) = \
+      STL_UTIL_UNIQUE(for_vector_vec).begin(); \
+  std::vector<TYPE*>::const_iterator STL_UTIL_UNIQUE(for_vector_end) = \
+      STL_UTIL_UNIQUE(for_vector_vec).end(); \
+  for (TYPE * VAL = NULL ; \
+       (VAL = ( STL_UTIL_UNIQUE(for_vector_iter) != \
+                STL_UTIL_UNIQUE(for_vector_end) ) ? \
+              * STL_UTIL_UNIQUE(for_vector_iter) : \
+              (TYPE*)0) ; \
+       STL_UTIL_UNIQUE(for_vector_iter)++ )
```

Comment:
This was already this way before your change, but pre-increment is preferred to post-increment where the result is not used, because in some compiler implementations post-increment creates a copy of the object.